### ISSUE TITLE
Thread cleanup

### ISSUE THREAD

**Issue by tamiko** [dealii/dealii#10391]
Dear all - just to be on the same page - I think we should quickly discuss what should happen regarding our Thread support cleanup.

I suggest to:

 - [ ] Remove direct thread handling from `SparseVanka` and port to `TaskGroup` (@drwells is working on this)
 - [ ] Port `step-28` to `TaskGroup` (instead of using threads directly)
 - [ ] Deprecate `ThreadGroup` and our `Thread` wrapper
 - [ ] Deprecate `Threads::Mutex` and use `std::mutex` directly. This will require some minor cleanup.
 - [ ] Remove all special handling of and `DEAL_II_WITH_THREADS` from the library.
 - [ ] I think that all special handling of `n_threads == 1` or `n_threads > 1` could go as well. (The`TaskGroup` should ideally work on immediately while blocking.)

**Comment by kronbichler** [comment]
One thing that will be a bit of work to port are the matrix-free loops here: https://github.com/dealii/dealii/blob/master/source/matrix_free/task_info.cc#L353-L585
That is essentially implementing a kind of task graph with TBB, but the code is super ugly. This should be much easier to implement with a DAG-based library like `taskflow` - but it needs to be done.

**Comment by tjhei** [comment]
> I suggest to:

I fully agree (we came up with this together on zoom).

To add to this:
ThreadGroup/Thread is not useful for any part inside the library as one should create a task instead. The only reason to spawn a thread is to do background I/O. Matthias points out that this doesn't need any fancy wrappers in the library.

**Comment by tjhei** [comment]
> but it needs to be done.

ah, okay. I will take a look at that as well.

**Comment by kronbichler** [comment]
I could help because the code is, well, involved, and I would need to somewhere sketch the dependencies. This is not super urgent, so if we run the stuff in serial for a week or two it won't hurt too much.

**Comment by tjhei** [comment]
> would need to somewhere sketch the dependencies.

That would be helpful. We can also video chat about it, if you think that helps.

**Comment by tamiko** [comment]
Some of our polynomials/fe classes use a `Thread::mutex` to guard scratch data. How shall I proceed? I can give all of them appropriate copy constructors/etc. or make the scratch area thread local (which would avoid contention)?

**Comment by bangerth** [comment]
`TaskGroup` should stay since we want to keep `Task`. `TaskGroup` is a useful concept .

`Threads::Mutex` is difficult to replace by `std::mutex` because the former is copyable whereas the latter is not. In essence, if you want to make the replacement, you will have to write lots of copy operators by hand that would otherwise be trivial.

**Comment by tamiko** [comment]
> Threads::Mutex is difficult to replace by std::mutex because the former is copyable whereas the latter is not. In essence, if you want to make the replacement, you will have to write lots of copy operators by hand that would otherwise be trivial.

It doesn't look too bad. Also, this is the perfect chance to remove most of the awkward mutexes that guard scratch data and use thread local storage instead.

**Comment by bangerth** [comment]
No, that's really not the right thing to do. If an operation in a protected region creates a different task that is run on the same thread, you will get corrupted data. If you use a mutex, you will get a deadlock. I much prefer the deadlock.

**Comment by tamiko** [comment]
@bangerth I don't understand. I think we are talking about different things here.

**Comment by tjhei** [comment]
@tamiko I don't understand https://github.com/dealii/dealii/blob/bbd65bc0527d84c18670ff7014e42aae1768ac77/cmake/config/CMakeLists.txt#L202
can you explain?

**Comment by tamiko** [comment]
@tjhei Yes. This line unconditionally sets `SET(DEAL_II_WITH_THREADS ON)` in the project configuration. If you want to have a different behavior for now you can simply replace the `ON` by `${DEAL_II_WITH_TBB}` (or whatever).

**Comment by tjhei** [comment]
> If you want to have a different behavior for now

We now always support creation of threads, so I don't see a situation where this would be necessary. Maybe we should just remove this define after #10418 is merged?

**Comment by tamiko** [comment]
On Fri, May 29, 2020, at 08:54 CDT, Timo Heister <notifications@github.com> wrote:

>> If you want to have a different behavior for now
>
> We now always support creation of threads, so I don't see a situation
> where this would be necessary. Maybe we should just remove this define
> after #10418 is merged?

Let's keep it there for backwards compatibility :-)

**Comment by tjhei** [comment]
I am looking for some initial feedback for an implementation of Workstream::run() using taskflow. I have made 4 different implementations so far and benchmark them against TBB. v4 seems to be more or less competitive on my laptop and my workstation (shown below).

Questions:
1. Right now the benchmark is hacked into step-9 and only considers the step-9 assembly. What other cases situations do we need to consider?
2. How do we handle the case if both taskflow and TBB are present in Workstream::run()? Does this need to be a user choice for now?
3. Any thoughts on the implementation or suggestions for improvement?
4. Any objections against putting v4 into the library? 

The implementations:
```
  // one task per cell, create scratch data and copydata as needed without reuse
  namespace taskflow_v1

  // Like v1, except that we do not create a big task graph, but we submit 512
  // jobs at a time.
  namespace taskflow_v2

  // Like v1 but work in chunks of size 8
  namespace taskflow_v3

  // Work in chunks of size 8 (only if we have enough items, otherwise 1) and
  // use a thread-local scratch object
  namespace taskflow_v4
```

Results (times and avg in seconds):
```
   Number of active cells:              16384
   Number of degrees of freedom:        410881
MultithreadInfo::n_cores()=64
** TASKFLOW v1 **
n_cores auto 1.51679 1.50997 1.52853 1.51086 1.51348  avg: 1.51593
n_cores 64 1.51102 1.5134 1.52004 1.50392 1.51464  avg: 1.5126
n_cores 32 1.57964 1.57532 1.55342 1.57089 1.58161  avg: 1.57218
n_cores 16 1.55728 1.56718 1.56371 1.55456 1.56536  avg: 1.56162
n_cores 8 1.56987 1.56361 1.55543 1.55586 1.55655  avg: 1.56026
n_cores 4 1.80923 1.80041 1.81217 1.81547 1.79817  avg: 1.80709
n_cores 2 2.72294 2.722 2.71871 2.71403 2.68753  avg: 2.71304
n_cores 1 1.8322 1.88404 1.83542 1.8941 1.83009  avg: 1.85517
** TASKFLOW v2 **
n_cores auto 1.63259 1.63799 1.62437 1.61467 1.62091  avg: 1.62611
n_cores 64 1.61539 1.63165 1.61086 1.61539 1.61741  avg: 1.61814
n_cores 32 1.63543 1.6166 1.62532 1.62592 1.63335  avg: 1.62733
n_cores 16 1.60035 1.59751 1.58873 1.59226 1.58611  avg: 1.59299
n_cores 8 1.57628 1.58537 1.58017 1.57572 1.5801  avg: 1.57953
n_cores 4 1.81652 1.82434 1.83171 1.83128 1.84647  avg: 1.83006
n_cores 2 2.72389 2.70127 2.73075 2.75308 2.73114  avg: 2.72803
n_cores 1 1.82033 1.81363 1.82171 1.82194 1.81696  avg: 1.81891
** TASKFLOW v3 **
n_cores auto 0.242558 0.253019 0.247955 0.249604 0.241506  avg: 0.246929
n_cores 64 0.247228 0.250037 0.254612 0.25079 0.258672  avg: 0.252268
n_cores 32 0.240371 0.238212 0.240435 0.246905 0.229582  avg: 0.239101
n_cores 16 0.239671 0.24713 0.254526 0.24723 0.24676  avg: 0.247064
n_cores 8 0.327448 0.337601 0.357137 0.343199 0.338321  avg: 0.340741
n_cores 4 0.601285 0.589835 0.584173 0.591084 0.58777  avg: 0.59083
n_cores 2 1.11473 1.11829 1.11961 1.11428 1.11829  avg: 1.11704
n_cores 1 1.83388 1.83794 1.84645 1.83717 1.84287  avg: 1.83966
** TASKFLOW v4 **
n_cores auto 0.187735 0.19767 0.184366 0.190977 0.190688  avg: 0.190287
n_cores 64 0.182609 0.190411 0.198562 0.206215 0.209467  avg: 0.197453
n_cores 32 0.152008 0.156053 0.160961 0.169807 0.163102  avg: 0.160386
n_cores 16 0.14974 0.148076 0.147577 0.141121 0.148324  avg: 0.146968
n_cores 8 0.232988 0.233688 0.232974 0.2329 0.233014  avg: 0.233113
n_cores 4 0.462408 0.459918 0.461322 0.464146 0.459715  avg: 0.461502
n_cores 2 0.909502 0.924066 0.923924 0.907118 0.912686  avg: 0.915459
n_cores 1 1.83354 1.83579 1.84175 1.83689 1.83954  avg: 1.8375
** TBB **
n_cores auto 0.24055 0.231847 0.238353 0.235988 0.235587  avg: 0.236465
n_cores 64 0.235965 0.23525 0.236473 0.236436 0.237271  avg: 0.236279
n_cores 32 0.19174 0.188839 0.187559 0.185877 0.184447  avg: 0.187692
n_cores 16 0.171227 0.170112 0.169096 0.16989 0.168672  avg: 0.1698
n_cores 8 0.231969 0.234809 0.235196 0.237732 0.234048  avg: 0.234751
n_cores 4 0.462603 0.460008 0.470515 0.468692 0.466051  avg: 0.465574
n_cores 2 0.928406 0.927998 0.932246 0.925651 0.921146  avg: 0.927089
n_cores 1 1.84179 1.84233 1.84535 1.84606 1.83977  avg: 1.84306
```

Conclusions:
- Optimtizations in v4 are necessary to beat TBB (which we do!)
- Hyperthreading is not beneficial on this machine, even worse: using 16 threads is fastest (this is an AMD EPYC 7542 32-Core Processor)

FYI: @bangerth 
see https://github.com/tjhei/dealii/tree/task-flow-test for the code.

**Comment by tamiko** [comment]
@tjhei The results for taskflow v4 look very promising.

Out of curiosity: Would you mind to run the assembly also for, say about 2-4M DoF (just comparing taskflow v4 versus TBB on say 16 cores). Maybe this shifts the sweet spot towards 32 threads - who knows.

It would be possible to look into this a bit with instrumentation via performance counters (e.g., looking at cache miss rates and false sharing) to try to understand the reason for the performance degradation a bit better. But I am not sure whether this is ultimately meaningful...

**Comment by bangerth** [comment]
Nice. I agree that it would be nice to see the same data for a larger problem. I can also easily try this on my 96-core (physical cores) machine one of these days!

**Comment by tjhei** [comment]
bigger run shows the same behavior:
```
   Number of active cells:              65536
   Number of degrees of freedom:        1640961
** TASKFLOW v4 **
n_cores auto 0.686414 0.722665 0.705674 0.699835 0.694059  avg: 0.701729
n_cores 64 0.675081 0.741168 0.728864 0.704213 0.756939  avg: 0.721253
n_cores 32 0.636024 0.663587 0.653752 0.598224 0.571377  avg: 0.624593
n_cores 16 0.594914 0.561403 0.559162 0.56769 0.56102  avg: 0.568838
n_cores 8 0.928786 0.922498 0.921857 0.924114 0.922296  avg: 0.92391
n_cores 4 1.83438 1.85249 1.83708 1.83845 1.83642  avg: 1.83976
n_cores 2 3.64288 3.64805 3.65075 3.65122 3.69915  avg: 3.65841
n_cores 1 7.33764 7.37859 7.33414 7.3271 7.57147  avg: 7.38979
** TBB **
n_cores auto 0.930327 0.911487 0.908379 0.90483 0.918538  avg: 0.914712
n_cores 64 0.915804 0.911007 0.912956 0.90962 0.908052  avg: 0.911488
n_cores 32 0.739986 0.716411 0.709554 0.700885 0.700279  avg: 0.713423
n_cores 16 0.680567 0.668818 0.666283 0.669826 0.667701  avg: 0.670639
n_cores 8 0.944075 0.939794 0.93654 0.933028 0.936142  avg: 0.937916
n_cores 4 1.88203 1.86659 1.87395 1.88551 1.89238  avg: 1.88009
n_cores 2 3.73691 3.72427 3.72942 3.74802 3.73694  avg: 3.73511
n_cores 1 7.32257 7.35453 7.35349 7.39146 7.32502  avg: 7.34941
```

> It would be possible to look into this a bit with instrumentation via performance counters

Yes, I agree that this is sensible. I would also want to move to a working version without TBB as soon as possible, though. Maybe put v4 in and go from there?

Should we preserve these different versions? I could convert this into a test.

**Comment by bangerth** [comment]
Yes, make a patch for `workstream.h` of v4 and I'll review. Make the other versions into a test just because you can and to give us confidence in TaskFlow going forward.

My take is that we want to get rid of TBB, so I don't see a need to keep the TBB version around in the long run. It may be interesting to keep it available as a backup for now, but would not object if you wanted to make TaskFlow the default now already.

**Comment by bangerth** [comment]
Then put the v4 into a namespace `WorkStream::v4` and when we get around to removing the old version, we just move stuff out of that namespace and into `WorkStream`.

**Comment by bangerth** [comment]
OK, so here's what I get on my 96 core-machine (x2 hyperthreaded) in debug mode, for refinement cycle 1 (16k cells, 410k unknowns):
![image](https://user-images.githubusercontent.com/7504421/84711504-c9a56f80-af23-11ea-9baa-50ae98ba6bbc.png)
So all of your versions scale about as well as the TBB version, with v4 slightly better than the TBB. All stop scaling beyond 24 cores.

In release mode, things look like this:
![image](https://user-images.githubusercontent.com/7504421/84711785-53553d00-af24-11ea-8a34-56e8b29e24e1.png)
Same conclusion (v4 as good or better than TBB), but this time scaling doesn't go beyond 12 cores.

**Comment by bangerth** [comment]
The separated point off to the right is the "automatic" choice the program outputs, which I believe means "no limit".

**Comment by bangerth** [comment]
And here are corresponding 3d results. Note that v4 scales substantially better than the TBB in release mode:
![image](https://user-images.githubusercontent.com/7504421/84726374-f8ced780-af49-11ea-94d5-936fec2bcff8.png)
This is using 64 cells and 9261 DoFs -- far too small to actually matter, but nice to see!

**Comment by bangerth** [comment]
For reference, I've added the data files from which these graphics are generated to the 9.3 release paper directory. I'm sure we'll want to write something about that.

**Comment by tjhei** [comment]
While not amazing, this is at least not worse than TBB.

> The separated point off to the right is the "automatic" choice the program outputs, which I believe means "no limit".

Yes, I am starting with the default setting without imposing a limit. This was mainly to test that our multithread-initialization is doing the right thing. :-)

Any thoughts for improvements? I would assume that the sequential copying is to blame for the limit in scaling, which depends on the cost of copying vs working. This is especially visible in your 3d example.

**Comment by bangerth** [comment]
Yes, precisely: This limits 2d scalability and is better in 3d because the work/copy ratio is better in 3d.

I think this data is pretty good, and would be happy to just replace the existing implementation by the one you have (v4). Or, if you have other examples you'd like me to test, I can do that too.

**Comment by tjhei** [comment]
colored version (PR coming soon) is working pretty well:
```
$ nice ./tests/multithreading/bench_01.release/bench_01.release 
   Number of active cells:              4096
   Number of degrees of freedom:        103041
MultithreadInfo::n_cores()=32 (limited from 32)
MultithreadInfo::n_threads()=32
0.459301 0.449928 0.459171  avg: 0.456134
build graph coloring time: 0.022838
** TASKFLOW v1 **
n_cores 32 0.384603 0.380422 0.384608  avg: 0.383211
n_cores 16 0.392884 0.378172 0.386371  avg: 0.385809
n_cores 8 0.369571 0.361329 0.366968  avg: 0.365956
n_cores 4 0.445397 0.444197 0.439561  avg: 0.443052
n_cores 2 0.661158 0.65769 0.646872  avg: 0.65524
n_cores 1 0.458006 0.454214 0.454295  avg: 0.455505
** TASKFLOW v4 **
n_cores 32 0.0690927 0.0617177 0.0592171  avg: 0.0633425
n_cores 16 0.0393967 0.0386836 0.0474251  avg: 0.0418351
n_cores 8 0.0598019 0.0600482 0.0596657  avg: 0.0598386
n_cores 4 0.117828 0.115915 0.116357  avg: 0.1167
n_cores 2 0.233914 0.229685 0.22814  avg: 0.23058
n_cores 1 0.457095 0.453487 0.454581  avg: 0.455055
** TBB **
n_cores 32 0.0575196 0.0558667 0.0547327  avg: 0.0560397
n_cores 16 0.0463896 0.0478459 0.046359  avg: 0.0468648
n_cores 8 0.0600865 0.0597485 0.0598673  avg: 0.0599008
n_cores 4 0.115311 0.115049 0.114824  avg: 0.115061
n_cores 2 0.230098 0.226304 0.230969  avg: 0.229124
n_cores 1 0.451154 0.449068 0.450612  avg: 0.450278
** TBB colored **
n_cores 32 0.0804186 0.0664614 0.0722916  avg: 0.0730572
n_cores 16 0.0538519 0.0541765 0.048578  avg: 0.0522021
n_cores 8 0.0668183 0.0676028 0.0656243  avg: 0.0666818
n_cores 4 0.121738 0.120167 0.120679  avg: 0.120861
n_cores 2 0.232952 0.229227 0.229783  avg: 0.230654
n_cores 1 0.47098 0.467761 0.466133  avg: 0.468291
** taskflow colored **
n_cores 32 0.0254161 0.0254552 0.0265732  avg: 0.0258148
n_cores 16 0.035147 0.0348996 0.0371388  avg: 0.0357284
n_cores 8 0.0624023 0.0624622 0.0626219  avg: 0.0624954
n_cores 4 0.120481 0.120122 0.119619  avg: 0.120074
n_cores 2 0.237503 0.233765 0.232194  avg: 0.234487
n_cores 1 0.472512 0.463799 0.467429  avg: 0.467914
```
(32 core AMD EPYC 7542, hyperthreading disabled)

**Comment by tamiko** [comment]
@drwells Agreed.

**Comment by bangerth** [comment]
@tamiko I think we agreed that TaskFlow is not, after all, the right direction. Should we just close this?

**Comment by bangerth** [comment]
@tamiko Ping?

**Comment by tjhei** [comment]
Well, I am unhappy with the current situation (buggy, new versions requiring c++17) and I am also not super motivated to invest more time into this right now. I do remember getting better performance even from the initial, simple taskflow implementation, though.

An alternative would be to import the newest taskflow (and disabling if we don't have c++17 support). We don't really need TBB and taskflow implementations at the same time. None of this would be in 9.4 of course.

Do you suggest we rip everything out and throw it away @bangerth ?

For 9.4 we should at least disable automatic detection of taskflow.

**Comment by bangerth** [comment]
I thought we *had* already ripped everything out. Did we not? Or did we only *decide* that we wanted to?

We should at least do so for the release. Nothing uses TaskFlow, there is no point shipping it as part of the release.

I think we had solutions for all places where we use the TBB right now with the exception of `parallel_for`. I do think that there is value in using something as widely distributed as the TBB, and be it only because that's what other projects do as well.

**Comment by tamiko** [comment]
@bangerth No we haven't ripped out everything yet. We did postpone the discussion the last time because @tjhei wasn't present :-)

As for taskflow - I think the safest option to proceed would be to simply remove the CMake taskflow bits on the release branch after we branched off. We have done so already for the last release.

**Comment by tamiko** [comment]
Another important thing to consider is what @kronbichler wrote about the matrix free thread support:
> One thing that will be a bit of work to port are the matrix-free loops here: https://github.com/dealii/dealii/blob/master/source/matrix_free/task_info.cc#L353-L585
That is essentially implementing a kind of task graph with TBB, but the code is super ugly. This should be much easier to implement with a DAG-based library like taskflow - but it needs to be done.

I want to point out that we lost thread support in matrix free for One API TBB at the moment. (The TBB API changed considerably and such a low-level dynamic task graph construction as we did in the matrix free implementation doesn't seem to be easily possible any more.)

**Comment by bangerth** [comment]
I don't know how TaskFlow and TBB compare for the execution of task graphs, but it has become quite simple to construct very powerful task graphs (including flowing data from one task to another) in new versions of the TBB. I quite liked the examples the documentation gives.

**Comment by tamiko** [comment]
Let's postpone the thread cleanup to 9.5 :-)
We will gut taskflow again for the release.

**Comment by bangerth** [comment]
I realized again that we still have the TaskFlow parts in the cmake configuration and in some setup routines. Should we nuke it? Or do we plan to go back to it once we require C++17?

**Comment by tjhei** [comment]
> I realized again that we still have the TaskFlow parts in the cmake configuration and in some setup routines. Should we nuke it? Or do we plan to go back to it once we require C++17?

My comment above still holds (taskflow has (had?) better performance but we need to require c++17 to use it).
Does it make sense to support TBB and taskflow? Not sure.

Maybe we need to have a group conversation about this topic.
